geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/data/RedisSortedSet.java

```diff
@@ -0,0 +1,286 @@
+    InternalDataSerializer.writePrimitiveInt(members.size(), out);
+    for (Map.Entry<byte[], byte[]> entry : members.entrySet()) {
+      byte[] key = entry.getKey();
+      byte[] value = entry.getValue();
+      InternalDataSerializer.writeByteArray(key, out);
+      InternalDataSerializer.writeByteArray(value, out);
+    }
+    InternalDataSerializer.writePrimitiveInt(sizeInBytes, out);
+  }
+
+  @Override
+  public void fromData(DataInput in, DeserializationContext context)
+      throws IOException, ClassNotFoundException {
+    super.fromData(in, context);
+    int size = InternalDataSerializer.readPrimitiveInt(in);
+    members = new Object2ObjectOpenCustomHashMap<>(size, ByteArrays.HASH_STRATEGY);
+    for (int i = 0; i < size; i++) {
+      members.put(InternalDataSerializer.readByteArray(in),
+          InternalDataSerializer.readByteArray(in));
+    }
+    sizeInBytes = InternalDataSerializer.readPrimitiveInt(in);
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (!(o instanceof RedisSortedSet)) {
+      return false;
+    }
+    if (!super.equals(o)) {
+      return false;
+    }
+    RedisSortedSet redisSortedSet = (RedisSortedSet) o;
+    if (members.size() != redisSortedSet.members.size()) {
+      return false;
+    }
+
+    for (Map.Entry<byte[], byte[]> entry : members.entrySet()) {
+      if (!Arrays.equals(redisSortedSet.members.get(entry.getKey()), (entry.getValue()))) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  @Override
+  public int getDSFID() {
+    return REDIS_SORTED_SET_ID;
+  }
+
+  protected synchronized AddOrChange memberAdd(byte[] memberToAdd, byte[] scoreToAdd) {
+    boolean added = (members.put(memberToAdd, scoreToAdd) == null);
+    if (added) {
+      sizeInBytes += calculateSizeOfNewFieldValuePair(memberToAdd, scoreToAdd);
+    }
+    return added ? AddOrChange.ADDED : AddOrChange.NOOP;
+  }
+
+  private synchronized void membersAddAll(AddsDeltaInfo addsDeltaInfo) {
+    Iterator<byte[]> iterator = addsDeltaInfo.getAdds().iterator();
+    while (iterator.hasNext()) {
+      byte[] member = iterator.next();
+      sizeInBytes += PER_PAIR_OVERHEAD + member.length;
+      byte[] score = iterator.next();
+      sizeInBytes += PER_PAIR_OVERHEAD + score.length;
+      members.put(member, score);
+    }
+  }
+
+
+  private synchronized void membersRemoveAll(RemsDeltaInfo remsDeltaInfo) {
+    for (byte[] member : remsDeltaInfo.getRemoves()) {
+      sizeInBytes -= (PER_PAIR_OVERHEAD + member.length);
+      members.remove(member);
+    }
+  }
+
+  /**
+   * @param region the region this instance is stored in
+   * @param key the name of the set to add to
+   * @param membersToAdd members to add to this set; NOTE this list may by modified by this call
+   * @return the number of members actually added
+   */
+  long zadd(Region<RedisKey, RedisData> region, RedisKey key, List<byte[]> membersToAdd,
```

Comment:
I think we could clean up this method a lot. We need to set the delta info whether or not it was changed or added (currently change is set as `NOOP` (from me, which will fall into the `default` in the case switch and set the delta to false).

We could do something like this:
```
  protected synchronized byte[] memberAdd(byte[] memberToAdd, byte[] scoreToAdd) {
    byte[] oldScore = members.put(memberToAdd, scoreToAdd);
    if (oldScore == null) {
      sizeInBytes += calculateSizeOfNewFieldValuePair(memberToAdd, scoreToAdd);
    } else {
      sizeInBytes += scoreToAdd.length - oldScore.length;
    }
    return oldScore;
  }
  
    long zadd(Region<RedisKey, RedisData> region, RedisKey key, List<byte[]> membersToAdd,
      SortedSetOptions options) {
    int membersAdded = 0;
    if (options == null) {
      options = new SortedSetOptions(SortedSetOptions.Exists.NONE, SortedSetOptions.Update.NONE);
    }

    Iterator<byte[]> iterator = membersToAdd.iterator();
    AddsDeltaInfo deltaInfo = null;
    while (iterator.hasNext()) {
      byte[] score = iterator.next();
      byte[] member = iterator.next();
      boolean newMember;
      if (options.isNX() && members.containsKey(member)) {
        continue;
      }
      if (options.isXX() && !members.containsKey(member)) {
        continue;
      }

      newMember = memberAdd(member, score) == null;

      // right now we don't need to worry about the changed option since we haven't added that yet...
      // we can change the if case once that's added.
      if (newMember) {
        membersAdded++;
      }

      // we need to add deltaInfo whether it was a new member or a changed member, we won't hit
      // this if we didn't add at all since we put `continue` for those cases
      // there did not seem to be a clean way to extract this to a method
      if (deltaInfo == null) {
        deltaInfo = new AddsDeltaInfo(new ArrayList<>());
      }
      deltaInfo.add(member);
      deltaInfo.add(score);
    }
    // store changes will check to make sure deltaInfo is not null
    storeChanges(region, key, deltaInfo);
    return membersAdded;
  }

Reply:
Refactored, pretty close to that code. Had to tweak the unit test slightly to provide SortedSetOptions arg.